File: Python/Construct Binary Tree from Inorder and Postorder Traversal.py
# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # @param inorder, a list of integers
    # @param postorder, a list of integers
    # @return a tree node
    def buildTree(self, inorder, postorder):
        len1 = len(inorder)
        if len1 == 0:
            return None
        index = 0
        for i in range(len1):
            if inorder[i] == postorder[-1]:
                index = i
                break
        root = TreeNode(postorder[-1])
        root.left = self.buildTree(inorder[:index], postorder[:index])
        root.right = self.buildTree(inorder[index+1:], postorder[index:len1-1])
        return root

File: C++/Construct-Binary-Tree-from-Inorder-and-Postorder-Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* helper(vector<int>& preorder, int preleft, int preright,
        vector<int>& inorder, int inleft, int inright) {
        if (preleft == preright) {
            TreeNode* root = new TreeNode(preorder[preleft]);
            return root;
        }
        if (preleft > preright) return NULL;
        int pos = -1;
        for (int i = inleft; i <= inright; i ++) {
            if (preorder[preleft] == inorder[i]) {
                pos = i;
                break;
            }
        }
        TreeNode* root = new TreeNode(preorder[preleft]);
        root->left = helper(preorder, preleft + 1, pos - inleft + preleft, inorder, inleft, pos - 1);
        root->right = helper(preorder, pos - inleft + preleft + 1, preright, inorder, pos + 1, inright);
        return root;
    }
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int len = preorder.size();
        return helper(preorder, 0, len - 1, inorder, 0, len - 1);
    }
};